app/models/article.rb

```diff
@@ -27,6 +27,7 @@ class Article < ApplicationRecord
 
   # The date that we began limiting the number of user mentions in an article.
   MAX_USER_MENTION_LIVE_AT = Time.utc(2021, 4, 7).freeze
+  IMG_MARKDOWN_REGEX = %r{!\[.*\]\(#{ApplicationConfig["APP_PROTOCOL"]}#{ApplicationConfig["AWS_BUCKET_NAME"]}.s3.amazonaws.com/#{ArticleImageUploader.new.store_dir}(.*)\)}.freeze # rubocop:disable Layout/LineLength
```

Comment:
Do we prescribe S3 usage or do we allow for any S3-compatible storage, such as DigitalOcean Spaces or MinIO? If we _only_ allow S3, this looks like it should work, but if we allow any S3-compatible object storage, we may not want to hardcode S3 here.

/cc @rhymes

Reply:
Seconded

Reply:
@jgaskins I think this PR is only intended to work on S3 and Cloudinary. I really don't know how other compatible storages can work as our entire configuration is based on S3:

https://github.com/forem/forem/blob/77885e8fa4c473f56ec20d273613ca751e22a78e/config/initializers/carrierwave.rb#L24

cc @maestromac

Reply:
_TL;DR: I'd like to get a lot more clarity on this from Product before we merge._

@rhymes DigitalOcean Spaces, MinIO, and at least a half-dozen other object-storage providers are API-compatible with S3. When I say API-compatible, I mean you could even use AWS's own S3 client to talk to them. The only thing you have to change is to provide the `endpoint` parameter:

<img width="937" alt="If you need to connect to a non-standard endpoint, you may specify the :endpoint option" src="https://user-images.githubusercontent.com/108205/125637582-2b9b374e-da91-4005-a098-73ee54627098.png">

<cite>https://docs.aws.amazon.com/sdk-for-ruby/v3/api/</cite>

It's literally that simple. Fog supports [this `:endpoint` option](https://github.com/fog/fog-aws/blob/1210edde0fbff8f508ac60ebcfc0c7ca6136d678/lib/fog/aws/storage.rb#L435-L451), as well. I don't know if we provide that, but if not it'd be trivial to add. However, it does mean we need to think about whether we're forcing people to use S3 specifically or giving options for alternative, API-compatible object storage. This can be important for folks hosting a Forem instance inside their own cluster since MinIO can run in any containerized environment.

All this may sound like it's outside the scope of this PR, but I assure you it's not. 🙂 If we assume everyone's using S3 then we are implicitly deciding they must use S3. I don't know of any product guidance one way or the other, so maybe we should talk about it with Product before we make a decision that I don't think is ours to make. If someone is currently using S3-compatible storage that isn't AWS S3 and this gets merged in, these requests may fail for them and cause problems.

Reply:
Thanks for the context! I do not know what's the answer here as this wasn't definitely on the radar when the issue was discussed and I'm not familiar about self host setups. Can you bring it to the attention of the Systems and Product team and follow up? This way @squarebat doesn't continue working on this PR before we have a final decision about what we want to do...